iked/dh.c

```diff
@@ -888,3 +916,123 @@ kemsx_create_shared2(struct dh_group *group, struct ibuf **sharedp,
 	*sharedp = buf;
 	return (0);
 }
+
+/* combine mlkem768 with curve25519 */
+int
+kemsx_init1(struct dh_group *group)
+{
+	/* delayed until kemsx_create_exchange3 */
+	return (0);
+}
+int
+kemsx_getlen1(struct dh_group *group)
+{
+	return (0);
+}
+int
+kemsx_create_exchange3(struct dh_group *group, struct ibuf **bufp,
+    struct ibuf *iexchange)
+{
+	struct kemsx_key1	*kemsx;
+	struct curve25519_key	*curve25519;
+	struct ibuf		*buf = NULL;
+	u_char *cp, *pk;
+	size_t have, need;
+	if (ec25519_init(group) == -1)
+		return (-1);
+	if (group->curve25519 == NULL)
+		return (-1);
+	if ((kemsx = calloc(1, sizeof(*kemsx))) == NULL)
+		return (-1);
+	group->kemsx = kemsx;
+	if (iexchange == NULL) {
+		kemsx->initiator = 1;
+		crypto_kem_mlkem768_keypair(kemsx->public, kemsx->secret);
+		/* output */
+		need = crypto_kem_mlkem768_PUBLICKEYBYTES +
+		    CURVE25519_SIZE;
+		buf = ibuf_new(NULL, need);
+		if (buf == NULL)
+			return -1;
+		cp = buf->buf;
+		memcpy(cp, kemsx->public,
+		    crypto_kem_mlkem768_PUBLICKEYBYTES);
+		cp += crypto_kem_mlkem768_PUBLICKEYBYTES;
+	} else {
+		kemsx->initiator = 0;
+		/* input */
+		have = ibuf_size(iexchange);
+		need = crypto_kem_mlkem768_PUBLICKEYBYTES +
+		    CURVE25519_SIZE;
+		if (have != need)
+			return -1;
+		/* output */
+		need = crypto_kem_mlkem768_CIPHERTEXTBYTES +
+		    CURVE25519_SIZE;
+		buf = ibuf_new(NULL, need);
+		if (buf == NULL)
+			return -1;
+		cp = buf->buf;
+		pk = iexchange->buf;
+		crypto_kem_mlkem768_enc(cp, kemsx->kemkey, pk);
+		cp += crypto_kem_mlkem768_CIPHERTEXTBYTES;
+	}
+	curve25519 = group->curve25519;
+	memcpy(cp, curve25519->public, CURVE25519_SIZE);
+	*bufp = buf;
+	return (0);
+}
+int
+kemsx_create_shared3(struct dh_group *group, struct ibuf **sharedp,
+    struct ibuf *exchange)
+{
+	struct curve25519_key	*curve25519 = group->curve25519;
+	struct kemsx_key1	*kemsx = group->kemsx;
+	struct ibuf		*buf = NULL;
+	EVP_MD_CTX		*ctx = NULL;
+	u_int8_t		*cp;
+	u_int8_t		 shared[CURVE25519_SIZE];
+	size_t			 have, need;
+	u_int			 len;
+	*sharedp = NULL;
+	if (kemsx == NULL)
+		return (-1);
+	if (exchange == NULL)
+		return (-1);
+	have = ibuf_size(exchange);
+	cp = exchange->buf;
+	if (kemsx->initiator) {
+		/* input */
+		need = crypto_kem_mlkem768_CIPHERTEXTBYTES +
+		    CURVE25519_SIZE;
+		if (have != need)
+			return (-1);
+		crypto_kem_mlkem768_dec(kemsx->kemkey, cp, kemsx->secret);
+		cp += crypto_kem_mlkem768_CIPHERTEXTBYTES;
+	} else {
+		/* input, should have been checked before */
+		need = crypto_kem_mlkem768_PUBLICKEYBYTES +
+		    CURVE25519_SIZE;
+		if (have != need)
+			return (-1);
+		cp += crypto_kem_mlkem768_PUBLICKEYBYTES;
+	}
+	crypto_scalarmult_curve25519(shared, curve25519->secret, cp);
+	/* result is hash of concatenation of KEM key and DH shared secret */
```

Comment:
I think in the case of ML-KEM we want to include the public keys in the hash too in order to achieve something close to MAL-BIND-K-PK security as defined in https://eprint.iacr.org/2023/1933.pdf.

https://durumcrustulum.com/2024/02/24/how-to-hold-kems/ provides a good summary.

Reply:
something like

        if (kemke->initiator) {
                /* we generate public key */
                kempub = kemke->public;
  ....
        } else {
                /* peer sends public key */
                kempub = cp;
...
        if ((ctx = EVP_MD_CTX_new()) == NULL ||
            EVP_DigestInit_ex(ctx, EVP_sha256(), NULL) != 1 ||
            EVP_DigestUpdate(ctx, kempub, ...) != 1 ||